lib/lib.bash

```diff
@@ -0,0 +1,363 @@
+#
+# Reports an error to stderr
+error() {
+  echo ERROR: "$@" >&2
+}
+
+# Usage:
+#   warn [MESSAGE...]
+#
+# MESSAGE: A message to output (works like echo)
+#
+# Reports a warning to stderr
+warn() {
+  echo WARNING: "$@" >&2
+}
+
+# Usage:
+#   header [MESSAGE...]
+#
+# MESSAGE: A message to output (works like echo)
+#
+# Outputs a header
+header() {
+  echo '~~~ :docker:' "$@"
+}
+
+# Usage:
+#   fail-build
+#
+# Function to hard fail, opening any collapsed output
+fail-build() {
+  echo "^^^ +++"
+  exit 1
+}
+
+# Usage:
+#   normalize-path PATH
+#
+# PATH: A POSIX path
+#
+# Normalizes a path: trailing slashes are removed, and multiple slashes are collapsed to a
+# single one.
+normalize-path() {
+  # The ${...} substition strips trailing slashes, and the -s flag to tr means to collapse
+  # multiple matches of the source set - in our case, we're lightly abusing it since the
+  # "replacement" is the same as the original character.
+  tr -s / / <<<"${1%%+(/)}"
+}
+
+# Usage:
+#   config-key OPTION
+#   config-key PLUGIN OPTION
+#
+# OPTION: A configuration option, in hyphenated format (e.g., "image-service").
+# PLUGIN: The name of a Buildkite plugin, which defaults to "extract". This exists to
+#         support peeking at the Docker Compose plugin's configuration in cases where we
+#         want to share it (mostly for the image-repository configuration option.)
+#
+# This function transforms kebab-cased option and plugin names into the corresponding
+# environment variable name, matching the format BUILDKITE_PLUGIN_<plugin name>_<option name>.
+config-key() {
+  local key
+  local plugin=extract
+  if test $# -eq 2; then
+    plugin="$1"
+    key="$2"
+  else
+    key="$1"
+  fi
+
+  echo "BUILDKITE_PLUGIN_$(env-format "$plugin")_$(env-format "$key")"
+}
+
+# Usage:
+#   get-config OPTION
+#   get-config PLUGIN OPTION
+#
+# See config-key for definition of PLUGIN and OPTION.
+#
+# This reads out the named configuration option. If the value is not present in the
+# environment, the empty string is returned.
+get-config() {
+  local key
+  key="$(config-key "$@")"
+
+  echo "${!key:-}"
+}
+
+# Usage:
+#   get-config-with-default OPTION DEFAULT
+#   get-config-with-default PLUGIN OPTION DEFAULT
+#
+# See config-key for definition of PLUGIN and OPTION.
+# DEFAULT: The default value to use instead of the empty string.
+#
+# This reads out the named configuration option. If the option is not present (or is
+# empty), then DEFAULT is returned instead.
+get-config-with-default() {
+  local key
+  local plugin=extract
+  local default
+
+  if test $# -eq 3; then
+    plugin="$1"
+    key="$2"
+    default="$3"
+  else
+    key="$1"
+    default="$2"
+  fi
+
+  local value
+  value="$(get-config "$plugin" "$key")"
+
+  echo "${value:-$default}"
+}
+
+# Usage:
+#   validate-config
+#
+# This function should be called first: it's crucial to ensure that the configuration is
+# fully validated before any actions are taken.
+validate-config() {
```

Comment:
Nit-pick: Breaking this up into a composition of several element-specific validation functions would make this much more readable and independently test-able. It would probably produce some additional helpful functions for reuse in similar buildkite plugins as well.

Reply:
There's a good conversation to be had there, but I'm going to leave this one for now. Let's take a closer look at our plugins together and see if there isn't a common library that can be factored out and included via, e.g., submodules or some other mechanism.